src/server.rs

```diff
@@ -58,65 +54,567 @@ use protocol::{
+        self.timeout = timeout;
+    }
+
+    /// Set the `force_recache` setting.
+    #[allow(dead_code)]
+    pub fn set_force_recache(&mut self, force_recache: bool) {
+        self.service.force_recache = force_recache;
+    }
+
+    /// Set the `force_recache` setting.
+    #[allow(dead_code)]
+    pub fn set_storage(&mut self, storage: Arc<Storage>) {
+        self.service.storage = storage;
+    }
+
+    /// Returns a reference to the thread pool this server will run work on
+    #[allow(dead_code)]
+    pub fn pool(&self) -> &CpuPool {
+        &self.service.pool
+    }
+
+    /// Returns a reference to the command creator this server will use
+    #[allow(dead_code)]
+    pub fn command_creator(&self) -> &C {
+        &self.service.creator
+    }
+
+    /// Returns the port that this server is bound to
+    #[allow(dead_code)]
+    pub fn port(&self) -> u16 {
+        self.listener.local_addr().unwrap().port()
+    }
+
+    /// Returns a handle to the event loop that this server is bound to.
+    #[allow(dead_code)]
+    pub fn handle(&self) -> Handle {
+        self.core.handle()
+    }
+
+    /// Runs this server to completion.
+    ///
+    /// If the `shutdown` future resolves then the server will be shut down,
+    /// otherwise the server may naturally shut down if it becomes idle for too
+    /// long anyway.
+    pub fn run<F>(self, shutdown: F) -> io::Result<()>
+        where F: Future,
+    {
+        self._run(Box::new(shutdown.then(|_| Ok(()))))
+    }
+
+    fn _run<'a>(self, shutdown: Box<Future<Item = (), Error = ()> + 'a>)
+                -> io::Result<()>
+    {
+        let SccacheServer { mut core, listener, rx, service, timeout, wait } = self;
+
+        // Create our "server future" which will simply handle all incoming
+        // connections in separate tasks.
+        let handle = core.handle();
+        let server = listener.incoming().for_each(move |(socket, _addr)| {
+            SccacheProto.bind_server(&handle, socket, service.clone());
+            Ok(())
+        });
+
+        // Right now there's a whole bunch of ways to shut down this server for
+        // various purposes. These include:
+        //
+        // 1. The `shutdown` future above.
+        // 2. An RPC indicating the server should shut down
+        // 3. A period of inactivity (no requests serviced)
+        //
+        // These are all encapsulated wih the future that we're creating below.
+        // The `ShutdownOrInactive` indicates the RPC or the period of
+        // inactivity, and this is then select'd with the `shutdown` future
+        // passed to this function.
+        let handle = core.handle();
+        let shutdown_idle = ShutdownOrInactive {
+            rx: rx,
+            timeout: Timeout::new(timeout, &handle)?,
+            handle: handle.clone(),
+            timeout_dur: timeout,
+        };
+
+        let server = server.select(shutdown_idle);
+        let shutdown = shutdown.map_err(|()| {
```

Comment:
Nah unfortunately error-chain isn't *great* with futures today but I've been looking forward to integrating it more into sccache, both error-chain itself and seeing if I can make the error propagation nicer (unsure of what that would look like)